2_process/src/pair_nhd_catchments.R

```diff
@@ -0,0 +1,124 @@
+    comid_cat <- get_upstream_nhd_catchments(nhd_lines,xwalk_table,prms_line)
+  }
+  
+  return(comid_cat)
+  
+}
+
+
+
+get_intersecting_nhdplus_catchments <- function(prms_line,prms_hrus,min_area_overlap){
+  #' @description This function uses a spatial join to find the NHDPlusV2 catchments that overlap the PRMS HRU's.
+  #' 
+  #' @param prms_line sf linestring containing a target PRMS segment
+  #' @param prms_hrus sf (multi)polygon containing the HRU's from the NHGF
+  #' @param min_area_overlap float; value indicating the minimum proportion of NHDPlusV2 catchment area that 
+  #' overlaps the PRMS polygon in order to be retained
+  #' 
+  
+  # Select corresponding HRU polygon 
+  prms_poly <- prms_hrus %>% 
+    filter(hru_segment == prms_line$subsegseg) %>% 
+    st_union()
+  
+  if(length(st_geometry(prms_poly))>0){
+    
+    # Download NHDPlusV2 catchments and calculate original area
+    nhd_cats <- nhdplusTools::get_nhdplus(AOI = prms_poly,realization = "catchment") %>%
+      suppressMessages() %>%
+      mutate(area_orig = as.numeric(st_area(.)))
+    
+    # Find intersection between NHD catchments and HRU polygon and calculate *intersecting* area
+    nhd_cats_intrsct <- st_intersection(nhd_cats,prms_poly) %>%
+      suppressWarnings() %>%
+      mutate(area_int = as.numeric(st_area(.)),
+             area_prop = as.numeric(area_int/area_orig)) %>%
+      # retain NHD catchments that overlap >= X% of their area with the HRU polygon
+      filter(area_prop >= min_area_overlap)
+    
+    # Save intersecting catchment COMID's
+    comids_all <- data.frame(PRMS_segid = prms_line$subsegid,
+                             comid_cat = paste(sort(unique(nhd_cats_intrsct$featureid)),collapse=";"))
+    
+  } else {
+    comids_all <- data.frame(PRMS_segid = prms_line$subsegid,
+                             comid_cat = NA)
+  }
+  
+  return(comids_all)
+  
+}
+
+
+
+get_upstream_nhd_catchments <- function(nhd_lines,xwalk_table,prms_line){
+  #' @description This function finds the NHDPlusV2 catchments that directly drain to a given segment 
+  #' by navigating the NHDPlusV2 value-added attributes tables using the information in the 
+  #' NHDPlus/PRMS reach-to-segment xwalk table
+  #' 
+  #' @param prms_line sf linestring containing a target PRMS segment
+  #' @param xwalk_table data frame containing the NHDPlusV2/PRMS reach-to-segment crosswalk information
+  #' @param nhd_lines sf object containing NHDPlusV2 flowlines for area of interest
+  #' nhd_lines must contain variables COMID, PATHLENGTH, LENGTHKM, and HYDROSEQ
+  #' 
+    
+  # subset xwalk for prms_line
+  xwalk_prms_line <- xwalk_table %>%
+    filter(PRMS_segid == prms_line$subsegid)
+  
+  # Find all tributaries upstream of comid_down
+  nhd_reach_down_UT <- nhdplusTools::get_UT(nhd_lines,xwalk_prms_line$comid_down)
+  
+  # Find GF POI's (=respective comid_down's) that are upstream of the PRMS segment of interest
```

Comment:
do you define GF elsewhere? Geospatial Fabric

Reply:
Maybe not (I guess I just leaned into the USGS-speak there). I can make that change in the revision to this PR, though.